native/cocos/2d/renderer/Batcher2d.cpp

```diff
@@ -0,0 +1,258 @@
+    _indexStart = currMeshBuffer->getIndexOffset();
+    //}
+
+    this->_currBID = -1;
+
+    // stencilstage
+
+    // Todo blendState & stencil State
+    auto* curdrawBatch = _drawBatchPool.alloc();
+    curdrawBatch->setVisFlags(_currLayer);
+    // curdrawBatch
+    curdrawBatch->setInputAssembler(ia);
+    curdrawBatch->setUseLocalFlag(nullptr); // todo usLocal
+    curdrawBatch->fillPass(_currMaterial, nullptr, 0, nullptr, 0);
+    auto& _pass = curdrawBatch->getPasses().at(0);
+
+    curdrawBatch->setDescriptorSet(getDescriptorSet(_currTexture, _currSampler, _pass->getLocalSetLayout()));
+    _batches.push_back(curdrawBatch);
+}
+
+void Batcher2d::resetRenderStates() {
+    _currMaterial = nullptr;
+    _currTexture = nullptr;
+    _currTextureHash = 0;
+    _currSampler = nullptr;
+    _currSamplerHash = 0;
+    _currLayer = 0;
+    _currEntity = nullptr;
+}
+
+gfx::DescriptorSet* Batcher2d::getDescriptorSet(gfx::Texture* texture, gfx::Sampler* sampler, gfx::DescriptorSetLayout* _dsLayout) {
+    ccstd::hash_t hash = 2;
+    if (texture != nullptr) {
+        ccstd::hash_combine(hash, texture->getHash());
+    }
+    if (sampler != nullptr) {
+        ccstd::hash_combine(hash, sampler->getHash());
+    }
+    auto iter = _descriptorSetCache.find(hash);
+    if (iter != _descriptorSetCache.end()) {
+        return iter->second;
+    }
+    _dsInfo.layout = _dsLayout;
+    auto ds = getDevice()->createDescriptorSet(_dsInfo);
+
+    ds->bindTexture(static_cast<uint>(pipeline::ModelLocalBindings::SAMPLER_SPRITE), texture);
+    ds->bindSampler(static_cast<uint>(pipeline::ModelLocalBindings::SAMPLER_SPRITE), sampler);
+    ds->update();
+
+    _descriptorSetCache.emplace(hash, ds);
+
+    return ds;
+}
+
+// update vertex code
+
+bool Batcher2d::initialize() {
+    return true;
+}
+
+void Batcher2d::update() {
+    updateVertDirtyRenderer();
+    fillBuffersAndMergeBatches();
+    resetRenderStates();
+
+    for (const auto scene : Root::getInstance()->getScenes()) {
+        for (const auto batch : _batches) {
+            scene->addBatch(batch);
+        }
+    }
+}
+
+void Batcher2d::updateVertDirtyRenderer() {
+    size_t size = _vertDirtyRenderers.size();
+    for (uint32_t i = 0; i < size; i++) {
+        fillVertexBuffers(_vertDirtyRenderers[i]);
+    }
+    _vertDirtyRenderers.clear();
+}
+
+void Batcher2d::uploadBuffers() {
+    if (_batches.size() > 0) {
+        //_meshDataArray.forEach(),uploadBuffers
+
+        for (index_t i = 0; i < _meshBuffers.size(); i++) {
+            UIMeshBuffer* buffer = _meshBuffers[i];
+            buffer->uploadBuffers();
+            buffer->reset();
+        }
+
+        updateDescriptorSet();
+    }
+}
+
+void Batcher2d::reset() {
+    for (index_t i = 0; i < _batches.size(); i++) {
+        scene::DrawBatch2D* batch = _batches[i];
+        batch->clear();
+        this->_drawBatchPool.free(batch);
+    }
+
+    // meshDataArray
+    for (index_t i = 0; i < _meshBuffers.size(); i++) {
```

Comment:
- please use for loop
- should clear _meshBuffers?

Reply:
done